File: dp/72.edit-distance.py
#
# @lc app=leetcode id=72 lang=python3
#
# [72] Edit Distance
#

# @lc code=start
class Solution:
    def minDistanceRecursive(self, word1: str, word2: str) -> int:
        # 返回 word1[0:i] 和 word2[0:j] 的最小编辑距离

        memo = {}

        def dp(i, j):
            if (i, j) in memo:
                return memo[(i, j)]

            if i == -1: return j + 1
            if j == -1: return i + 1

            if word1[i] == word2[j]:
                memo[(i, j)] = dp(i - 1, j - 1) # 什么也不做
            else:
                memo[(i, j)] = min(
                    dp(i, j - 1) + 1,   # 插入
                    dp(i - 1, j) + 1,   # 删除
                    dp(i - 1, j - 1) + 1    # 替换
                )
            return memo[(i, j)]

        return dp(len(word1) - 1, len(word2) - 1)

    def minDistance(self, word1: str, word2: str) -> int:
        m = len(word1)
        n = len(word2)
        dp = [[float('inf') for _ in range(m + 1)] for _ in range(n + 1)]
        dp[0] = [i for i in range(m + 1)]
        for i in range(n + 1):
            dp[i][0] = i

        for i in range(1, n + 1):
            for j in range(1, m + 1):
                if word2[i - 1] == word1[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = min(
                        dp[i - 1][j] + 1,
                        dp[i][j - 1] + 1,
                        dp[i - 1][j - 1] + 1,
                    )

        return dp[n][m]

# @lc code=end

File: dp/72.edit-distance.go
package leetcode

// minDistance TODO
// 72. 编辑距离
// DP
// 状态转移方程：D[i][j] = min(D[i][j-1]+1, D[i-1][j]+1, D[i-1][j-1]/D[i-1][j-1]+1)
//
//	add on A		add on B			modify on A
func minDistance(word1 string, word2 string) int {
	n, m := len(word1), len(word2)
	// 有一个字符串为空串
	if n*m == 0 {
		return n + m
	}
	D := make([][]int, n+1)
	for i := range D {
		D[i] = make([]int, m+1)
	}

	// init
	for i := range D {
		D[i][0] = i
	}
	for j := 0; j < m+1; j++ {
		D[0][j] = j
	}

	// DP
	for i := 1; i < n+1; i++ {
		for j := 1; j < m+1; j++ {
			left := D[i-1][j] + 1
			down := D[i][j-1] + 1
			leftDown := D[i-1][j-1]
			if word1[i-1] != word2[j-1] {
				leftDown += 1
			}
			D[i][j] = min(left, min(down, leftDown))
		}
	}
	return D[n][m]
}
